core/src/main/java/com/linecorp/armeria/server/DefaultDecodedHttpRequest.java

```diff
@@ -0,0 +1,160 @@
+import com.linecorp.armeria.common.RequestHeaders;
+import com.linecorp.armeria.internal.common.DefaultHttpRequest;
+import com.linecorp.armeria.internal.common.InboundTrafficController;
+
+import io.netty.channel.EventLoop;
+
+final class DefaultDecodedHttpRequest extends DefaultHttpRequest implements DecodedHttpRequest {
+
+    private final EventLoop eventLoop;
+    private final int id;
+    private final int streamId;
+    private final boolean keepAlive;
+    private final InboundTrafficController inboundTrafficController;
+    private final long maxRequestLength;
+    @Nullable
+    private ServiceRequestContext ctx;
+    private long transferredBytes;
+
+    @Nullable
+    private HttpResponse response;
+    private boolean isResponseAborted;
+
+    DefaultDecodedHttpRequest(EventLoop eventLoop, int id, int streamId, RequestHeaders headers,
+                              boolean keepAlive, InboundTrafficController inboundTrafficController,
+                              long maxRequestLength) {
+        super(headers);
+
+        this.eventLoop = eventLoop;
+        this.id = id;
+        this.streamId = streamId;
+        this.keepAlive = keepAlive;
+        this.inboundTrafficController = inboundTrafficController;
+        this.maxRequestLength = maxRequestLength;
+    }
+
+    @Override
+    public void init(ServiceRequestContext ctx) {
+        this.ctx = ctx;
+    }
+
+    @Override
+    public int id() {
+        return id;
+    }
+
+    @Override
+    public int streamId() {
+        return streamId;
+    }
+
+    @Override
+    public boolean isKeepAlive() {
+        return keepAlive;
+    }
+
+    long maxRequestLength() {
+        return ctx != null ? ctx.maxRequestLength() : maxRequestLength;
+    }
+
+    long transferredBytes() {
+        return transferredBytes;
+    }
+
+    void increaseTransferredBytes(long delta) {
+        if (transferredBytes > Long.MAX_VALUE - delta) {
+            transferredBytes = Long.MAX_VALUE;
+        } else {
+            transferredBytes += delta;
+        }
+    }
+
+    @Override
+    public EventLoop defaultSubscriberExecutor() {
+        return eventLoop;
+    }
+
+    @Override
+    public boolean tryWrite(HttpObject obj) {
+        assert ctx != null : "uninitialized DecodedHttpRequest must be aborted.";
+
+        final boolean published;
+        if (obj instanceof HttpHeaders) { // HTTP trailers.
+            published = super.tryWrite(obj);
+            ctx.logBuilder().requestTrailers((HttpHeaders) obj);
+            // Close this stream because HTTP trailers is the last element of the request.
+            close();
+        } else {
+            final HttpData httpData = (HttpData) obj;
+            httpData.touch(ctx);
+            published = super.tryWrite(httpData);
+            if (published) {
+                inboundTrafficController.inc(httpData.length());
+            }
+            ctx.logBuilder().increaseRequestLength(httpData);
```

Comment:
Would it be worth close this request early when `httpData.isEndOfStream()` is `true`?

Reply:
Sounds good.